Crash when running plan on VMs with multiple network interfaces with no gateway set

Hello:

I am running into a terraform crash that I can't seem to pin down or work around. I am creating VMs in vsphere. The VMs have two network interfaces attached one of which has no gateway defined. Creating the VM initially is no problem, but running subsequent `terraform plan` commands are producing terraform crashes. The crash.log is here:

https://gist.github.com/icksa/a87c295aa976bcd805981e2be16c03b8#file-crash-log

There terraform file is here:

https://gist.github.com/icksa/a87c295aa976bcd805981e2be16c03b8#file-terraform-tf

The relevant section of the crash logs appears to be:
```
2017-09-21T09:28:16.960-0500 [DEBUG] plugin.terraform-provider-vsphere_v0.3.0_x4: 2017/09/21 09:28:16 [DEBUG] networkInterface: map[string]interface {}{"key":4000, "ipv4_address":"10.200.96.51", "ipv4_prefix_length":20, "label":"Stratus6", "mac_address":"00:50:56:9a:12:f0"}
2017-09-21T09:28:16.960-0500 [DEBUG] plugin.terraform-provider-vsphere_v0.3.0_x4: 2017/09/21 09:28:16 [DEBUG] networks: []map[string]interface {}{map[string]interface {}{"ipv4_address":"10.200.96.51", "ipv4_prefix_length":20, "label":"Stratus6", "mac_address":"00:50:56:9a:12:f0", "key":4000}, map[string]interface {}{"label":"Stratus3", "mac_address":"00:50:56:9a:a0:5e", "key":4001, "ipv4_address":"10.200.20.71", "ipv4_prefix_length":22}}
2017-09-21T09:28:16.960-0500 [DEBUG] plugin.terraform-provider-vsphere_v0.3.0_x4: 2017/09/21 09:28:16 [DEBUG] ipv4_gateway of device id 5: 10.200.20.5
2017-09-21T09:28:16.963-0500 [DEBUG] plugin.terraform-provider-vsphere_v0.3.0_x4: panic: runtime error: index out of range
2017-09-21T09:28:16.963-0500 [DEBUG] plugin.terraform-provider-vsphere_v0.3.0_x4:
2017-09-21T09:28:16.963-0500 [DEBUG] plugin.terraform-provider-vsphere_v0.3.0_x4: goroutine 22 [running]:
2017-09-21T09:28:16.963-0500 [DEBUG] plugin.terraform-provider-vsphere_v0.3.0_x4: github.com/terraform-providers/terraform-provider-vsphere/vsphere.resourceVSphereVirtualMachineRead(0x4233d7570, 0x1e8b7e0, 0x42337a8d0, 0x1, 0x29d3aa0)
2017-09-21T09:28:16.963-0500 [DEBUG] plugin.terraform-provider-vsphere_v0.3.0_x4: 	/opt/teamcity-agent/work/222ea50a1b4f75f4/src/github.com/terraform-providers/terraform-provider-vsphere/vsphere/resource_vsphere_virtual_machine.go:1180 +0x46a7
2017-09-21T09:28:16.963-0500 [DEBUG] plugin.terraform-provider-vsphere_v0.3.0_x4: github.com/terraform-providers/terraform-provider-vsphere/vendor/github.com/hashicorp/terraform/helper/schema.(*Resource).Refresh(0x423467740, 0x4233ea370, 0x1e8b7e0, 0x42337a8d0, 0x423387508, 0x423469201, 0x80000000018)
2017-09-21T09:28:16.963-0500 [DEBUG] plugin.terraform-provider-vsphere_v0.3.0_x4: 	/opt/teamcity-agent/work/222ea50a1b4f75f4/src/github.com/terraform-providers/terraform-provider-vsphere/vendor/github.com/hashicorp/terraform/helper/schema/resource.go:321 +0x199
2017-09-21T09:28:16.963-0500 [DEBUG] plugin.terraform-provider-vsphere_v0.3.0_x4: github.com/terraform-providers/terraform-provider-vsphere/vendor/github.com/hashicorp/terraform/helper/schema.(*Provider).Refresh(0x423207e30, 0x4233ea320, 0x4233ea370, 0x8422fb4d90, 0x0, 0x18)
2017-09-21T09:28:16.963-0500 [DEBUG] plugin.terraform-provider-vsphere_v0.3.0_x4: 	/opt/teamcity-agent/work/222ea50a1b4f75f4/src/github.com/terraform-providers/terraform-provider-vsphere/vendor/github.com/hashicorp/terraform/helper/schema/provider.go:284 +0x9a
2017-09-21T09:28:16.963-0500 [DEBUG] plugin.terraform-provider-vsphere_v0.3.0_x4: github.com/terraform-providers/terraform-provider-vsphere/vendor/github.com/hashicorp/terraform/plugin.(*ResourceProviderServer).Refresh(0x423351a20, 0x42311cde0, 0x42311d050, 0x0, 0x0)
2017-09-21T09:28:16.963-0500 [DEBUG] plugin.terraform-provider-vsphere_v0.3.0_x4: 	/opt/teamcity-agent/work/222ea50a1b4f75f4/src/github.com/terraform-providers/terraform-provider-vsphere/vendor/github.com/hashicorp/terraform/plugin/resource_provider.go:510 +0x4e
2017-09-21T09:28:16.963-0500 [DEBUG] plugin.terraform-provider-vsphere_v0.3.0_x4: reflect.Value.call(0x42334f200, 0x4233ec640, 0x13, 0x1e9e540, 0x4, 0x422ffcf20, 0x3, 0x3, 0x0, 0x0, ...)
2017-09-21T09:28:16.963-0500 [DEBUG] plugin.terraform-provider-vsphere_v0.3.0_x4: 	/usr/local/go/src/reflect/value.go:434 +0x906
2017-09-21T09:28:16.963-0500 [DEBUG] plugin.terraform-provider-vsphere_v0.3.0_x4: reflect.Value.Call(0x42334f200, 0x4233ec640, 0x13, 0x4233e6f20, 0x3, 0x3, 0x423307020, 0x0, 0x0)
2017-09-21T09:28:16.963-0500 [DEBUG] plugin.terraform-provider-vsphere_v0.3.0_x4: 	/usr/local/go/src/reflect/value.go:302 +0xa4
2017-09-21T09:28:16.963-0500 [DEBUG] plugin.terraform-provider-vsphere_v0.3.0_x4: net/rpc.(*service).call(0x4233aadc0, 0x4234d5400, 0x4233d4710, 0x4233dcc80, 0x423351f80, 0x1b635c0, 0x42311cde0, 0x16, 0x1b63600, 0x42311d050, ...)
2017-09-21T09:28:16.963-0500 [DEBUG] plugin.terraform-provider-vsphere_v0.3.0_x4: 	/usr/local/go/src/net/rpc/server.go:381 +0x142
2017-09-21T09:28:16.963-0500 [DEBUG] plugin.terraform-provider-vsphere_v0.3.0_x4: created by net/rpc.(*Server).ServeCodec
2017-09-21T09:28:16.963-0500 [DEBUG] plugin.terraform-provider-vsphere_v0.3.0_x4: 	/usr/local/go/src/net/rpc/server.go:475 +0x36b
2017/09/21 09:28:16 [ERROR] root: eval: *terraform.EvalRefresh, err: vsphere_virtual_machine.atl-ch3-vm-mesosworker01: unexpected EOF
2017/09/21 09:28:16 [ERROR] root: eval: *terraform.EvalSequence, err: vsphere_virtual_machine.atl-ch3-vm-mesosworker01: unexpected EOF
2017/09/21 09:28:16 [TRACE] [walkRefresh] Exiting eval tree: vsphere_virtual_machine.atl-ch3-vm-mesosworker01
2017/09/21 09:28:16 [TRACE] dag/walk: upstream errored, not walking "provider.vsphere (close)"
2017-09-21T09:28:16.965-0500 [DEBUG] plugin: plugin process exited: path=/Users/milad/git/atlas-terraform/atlas-prod/ch3/vms/mesos-worker/.terraform/plugins/darwin_amd64/terraform-provider-vsphere_v0.3.0_x4
2017/09/21 09:28:16 [DEBUG] plugin: waiting for all plugin processes to complete...
2017-09-21T09:28:16.965-0500 [WARN ] plugin: error closing client during Kill: err="connection is shut down"
```

I am not sure exactly what is causing this but I would greatly appreciate any help. This crash is happening pretty regularly now...Please let me know what additional information to send.

Thanks

--- a/vsphere/resource_vsphere_virtual_machine.go
+++ b/vsphere/resource_vsphere_virtual_machine.go
@@ -1199,16 +1199,39 @@ func resourceVSphereVirtualMachineRead(d *schema.ResourceData, meta interface{})
 							gatewaySetting = "ipv4_gateway"
 						}
 						if gatewaySetting != "" {
-							deviceID, err := strconv.Atoi(route.Gateway.Device)
-							if len(networkInterfaces) == 1 {
-								deviceID = 0
-							}
+							// We need to get the correct device index which is not always
+							// aligned with what we have saved in Terraform state.
+							//
+							// The index in Device corresponds with the device index in
+							// Guest.Net, which we can use to discover the key.
+							did := route.Gateway.Device
+							nidx, err := strconv.Atoi(did)
 							if err != nil {
-								log.Printf("[WARN] error at processing %s of device id %#v: %#v", gatewaySetting, route.Gateway.Device, err)
-							} else {
-								log.Printf("[DEBUG] %s of device id %d: %s", gatewaySetting, deviceID, route.Gateway.IpAddress)
-								networkInterfaces[deviceID][gatewaySetting] = route.Gateway.IpAddress
+								// "device" for some reason is a string, even though its result
+								// has always been observed to be an int. This should never
+								// happen, but we warn just in case.
+								log.Printf("[WARN] error at processing %s of device id %#v: %#v", gatewaySetting, did, err)
+								continue
 							}
+							key := mvm.Guest.Net[nidx].DeviceConfigId
+							if key < 0 {
+								// Gateway is not set to any hardware device, ignore.
+								continue
+							}
+							didx := -1
+							for n, dev := range networkInterfaces {
+								if dev["key"] == key {
+									didx = n
+									break
+								}
+							}
+							if didx < 0 {
+								// We are not tracking the device that the gateway is going to
+								// in state, so just skip the gateway.
+								continue
+							}
+							log.Printf("[DEBUG] %s of device id %d: %s", gatewaySetting, didx, route.Gateway.IpAddress)
+							networkInterfaces[didx][gatewaySetting] = route.Gateway.IpAddress
 						}
 					}
 				}

